### ISSUE TITLE
0.22.0 introducing breaking Generics in "post", "put" and "patch"

### ISSUE THREAD

**Issue by iwt-gregorpoloczek** [axios/axios#4130]
Description:
The signature of the following methods changed from (0.21.4)
```typescript
  post<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R>;
  put<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R>;
  patch<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R>;
```
to (0.22.0)
```typescript
  post<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
  put<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
  patch<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;
```
breaking every usage of the method, when using generics, e.g.:
```typescript
type RequestData = {
  input: string;
};

type ResponseData = {
  value: string;
};

async function test(): Promise<void> {
  const requestData: RequestData = { input: "input" }
  const {data} = await axios.post<ResponseData>("https://some-endpoint.com/v1/endpoint", requestData);
  console.log(data.value);
}
```
The new generics will now cause compile errors, since ``RequestData`` and ``ResponseData`` are not the same type.

The following workaround helps:
```typescript
async function test(): Promise<void> {
  const requestData: RequestData = { input: "input" }
  const {data} = await axios.post<RequestData, AxiosReponse<ResponseData>>("https://some-endpoint.com/v1/endpoint", requestData);
  console.log(data.value);
}
```

These changes seem arbitrary, and possibly done by mistake, as it is very unlikely, that a Post-Request has **exactly** the same type for its payload, as it has for its response.

**Comment by gstamac** [comment]
You were a bit quicker then me :) https://github.com/axios/axios/issues/4131

**Comment by jamesots-tq** [comment]
Looks like there's a PR open for this at the moment: #4111

**Comment by iwt-gregorpoloczek** [comment]
> Looks like there's a PR open for this at the moment: #4111

Thanks for the info!

**Comment by TheVaan** [comment]
Duplicate to #4109

**Comment by iwt-gregorpoloczek** [comment]
Closed because of duplication #4109 thanks @TheVaan.
